Adding the IFEval scenario

liamjxu opened the pull request:
Adding the scenario, run specs, and metric for [IFEval](https://arxiv.org/pdf/2311.07911). No new adapters were added, instead reused the `GenerationAdapter`.

yifanmai commented on src/helm/benchmark/metrics/ifeval_instructions.py:
1. Move these files to a `ifeval` subpackage within `metrics` (except `ifeval_metrics.py` which can stay under `metrics`).
2. Don't run the linter on this file; just reproduce the raw contents exactly (except for import statements). This makes it easier for someone to audit that the code is unchanged using `diff`.
3. Add the following lines to the start of the file to skip the linter:

```
# flake8: noqa
# type: ignore
# The following code has reproduced with minor modifications to `import` statements from the following URL:
# https://github.com/google-research/google-research/blob/c7f60c013623e613732a096e2a0c2872491ec912/instruction_following_eval/instructions.py
```

Tip: you can get the [permalink version](https://github.com/google-research/google-research/blob/c7f60c013623e613732a096e2a0c2872491ec912/instruction_following_eval/instructions.py) of the GitHub URL with the githash by going to the [latest version](https://github.com/google-research/google-research/blob/master/instruction_following_eval/instructions.py) and pressing 'y' on your keyboard.

Likewise for the other `ifeval_instructions*` files.

liamjxu replied:
Addressed in the latest change.

P.S. Thanks for the tip, it's convenient! The shortcut for me is `Shift+Ctrl+,` for some unknown reason, but it works too.

yifanmai replied:
Might be an OS / browser specific thing.

yifanmai commented on src/helm/benchmark/metrics/ifeval_metrics.py:
```diff
@@ -0,0 +1,44 @@
+from helm.benchmark.metrics.metric_service import MetricService
+from helm.benchmark.metrics.statistic import Stat
+
+import src.helm.benchmark.metrics.ifeval_instructions_registry as instructions_registry
```
Remove `src.` from this import.

liamjxu replied:
Addressed in the latest change.

yifanmai commented on src/helm/benchmark/metrics/ifeval_instructions.py:
```diff
@@ -0,0 +1,1500 @@
+from absl import logging
+import langdetect
+
+import src.helm.benchmark.metrics.ifeval_instructions_util as instructions_util
+
+from src.helm.benchmark.metrics.ifeval_instructions_util import LANGUAGE_CODES
```
Delete `src.` from these imports.

Likewise for the other files.

liamjxu replied:
Addressed in the latest change.

yifanmai commented on src/helm/benchmark/metrics/ifeval_metrics.py:
```diff
@@ -0,0 +1,44 @@
+        response = request_state.result.completions[0].text.strip()
+
+        is_following_list = []
+        for index, instruction_id in enumerate(instruction_id_list):
```
Add a link in a comment to https://github.com/google-research/google-research/blob/c7f60c013623e613732a096e2a0c2872491ec912/instruction_following_eval/evaluation_main.py#L96-L125

liamjxu replied:
Addressed in the latest change.

yifanmai commented on src/helm/benchmark/metrics/ifeval_metrics.py:
```diff
@@ -0,0 +1,44 @@
+            else:
+                is_following_list.append(0)
+
+        return [Stat(MetricName("strict_accuracy")).add(sum(is_following_list) / len(is_following_list))]
```
nit: `"strict_accuracy"` -> `"ifeval_strict_accuracy"` - the name is generic enough that we should probably namespace it.

liamjxu replied:
Addressed in the latest change.

yifanmai reviewed (changes requested):
Awesome, thanks! Left some comments.

Could you open a separate pull request to add IFEval to `schema_lite_v2.yaml`? Basically, follow what has already been done with `gpqa`. You'll also need to add `ifeval_strict_accuracy` to the metrics (follow `exact_match`).

liamjxu commented:
> Also I have no idea why the tests are failing... I'll look into it.

I think this is because the IFEval code has dependencies not in helm yet. I installed them locally to make it run.

liamjxu commented:
> Awesome, thanks! Left some comments.
> 
> Could you open a separate pull request to add IFEval to `schema_lite_v2.yaml`? Basically, follow what has already been done with `gpqa`. You'll also need to add `ifeval_strict_accuracy` to the metrics (follow `exact_match`).

Sure, I will do this after addressing all the comments.

liamjxu commented on src/helm/benchmark/scenarios/ifeval_scenario.py:
```diff
@@ -0,0 +1,45 @@
+        super().__init__()
+
+    def get_instances(self, output_path: str) -> List[Instance]:
+        # Get GPQA from HuggingFace
```
Addressed in the latest change.

yifanmai replied:
Still seems to be here?

liamjxu commented on src/helm/benchmark/scenarios/ifeval_scenario.py:
```diff
@@ -0,0 +1,45 @@
+
+    name = "ifeval"
+    description = "Instruction-Following Evaluation for Large Language Models"
+    tags = ["question answering"]
```
Addressed in the latest change.

liamjxu commented on src/helm/benchmark/metrics/common_metric_specs.py:
```diff
@@ -165,3 +165,7 @@ def get_disinformation_metric_specs(args: Optional[Dict] = None) -> List[MetricS
+
+
+def get_ifeval_metric_specs() -> List[MetricSpec]:
+    return [MetricSpec(class_name="helm.benchmark.metrics.ifeval_metrics.IFEvalMetric")]
```
Got it, thanks!

I think you meant `metric_specs = ...`, not adapter?

With that assumption, addressed in the latest change.

liamjxu commented:
```
Found 12 errors in 2 files (checked 650 source files)
src/helm/benchmark/scenarios/test_ifeval_scenario.py:31: error: Invalid index type "str" for "str | Any"; expected type "SupportsIndex | slice"  [index]
src/helm/benchmark/scenarios/test_ifeval_scenario.py:32: error: Invalid index type "str | Any" for "str | Any"; expected type "SupportsIndex | slice"  [index]
src/helm/benchmark/scenarios/test_ifeval_scenario.py:33: error: Invalid index type "str" for "str | Any"; expected type "SupportsIndex | slice"  [index]
src/helm/benchmark/scenarios/test_ifeval_scenario.py:34: error: Invalid index type "str" for "str | Any"; expected type "SupportsIndex | slice"  [index]
src/helm/benchmark/scenarios/test_ifeval_scenario.py:35: error: Invalid index type "str | Any" for "str | Any"; expected type "SupportsIndex | slice"  [index]
src/helm/benchmark/metrics/ifeval_metrics.py:22: error: Value of type "dict[str, str] | None" is not indexable  [index]
src/helm/benchmark/metrics/ifeval_metrics.py:23: error: Value of type "dict[str, str] | None" is not indexable  [index]
src/helm/benchmark/metrics/ifeval_metrics.py:34: error: Module has no attribute "INSTRUCTION_DICT"  [attr-defined]
src/helm/benchmark/metrics/ifeval_metrics.py:37: error: Item "str" of "str | Any" has no attribute "items"  [union-attr]
Error: Process completed with exit code 1.
```

I looked into the testing failure errors and realized that the type checker was failing. 

In the current implementation, the `extra_data` field in the `Instance` class is annotated with type `Optional[Dict[str, str]]`, yet in IFEval, `instruction_ids` maps to a list of strings and `instruction_kwargs` maps to a list of dictionaries.

Should we linearize IFEval's extra data, or should we update the type annotation of the `extra_data` field?

yifanmai commented:
Let's change `extra_data` to type `Dict[str, Any]` - does this make the type checker pass? My opinion is that the value can consist of any JSON serializable object, including nested dicts and lists,

liamjxu commented:
@yifanmai This PR is ready for review for merging

yifanmai commented on src/helm/benchmark/static/schema_lite_v2.yaml:
```diff
@@ -128,12 +133,30 @@ run_groups:
       - gpqa
+      - ifeval
+
+  # - name: mmlu_pro
+  #   display_name: MMLU-Pro
+  #   description: MMLU-Pro
+  #   metric_groups:
+  #     - accuracy
+  #     - efficiency
+  #     - general_information
+  #   environment:
+  #     main_name: exact_match  # non-CoT
+  #     main_split: test
+  #   taxonomy:
+  #     task: "?"
+  #     what: "?"
+  #     who: "?"
+  #     when: "?"
+  #     language: English
```
Revert this change (otherwise it will conflict with the MMLU-Pro pull request)

liamjxu replied:
Makes sense. Both comments are addressed in the latest change.

yifanmai reviewed (approved):
LGTM
